File: lib/agent/site_supervisor.dart
import 'dart:async';
import 'dart:convert';

import 'package:vault/agent/site_supervisor_script.dart';
import 'package:vault/sandbox/proot_workspace.dart';
import 'package:vault/sandbox/sandbox_models.dart';
import 'package:vault/sandbox/wsl_workspace.dart';

export 'package:vault/agent/site_supervisor_script.dart';

/// One line from the supervisor event log.
class SiteSupervisorEvent {
  const SiteSupervisorEvent({
    required this.id,
    required this.state,
    this.pid,
    this.reason,
  });

  final String id;
  final String state;
  final int? pid;
  final String? reason;

  bool get listening => state == 'listening';
  bool get down =>
      state == 'exited' || state == 'port_lost' || state == 'start_failed';

  static SiteSupervisorEvent? tryParse(String line) {
    final trimmed = line.trim();
    if (trimmed.isEmpty) return null;
    try {
      final json = jsonDecode(trimmed);
      if (json is! Map) return null;
      final id = (json['id'] as String?)?.trim() ?? '';
      final state = (json['state'] as String?)?.trim() ?? '';
      if (id.isEmpty || state.isEmpty) return null;
      return SiteSupervisorEvent(
        id: id,
        state: state,
        pid: json['pid'] is int ? json['pid'] as int : int.tryParse('${json['pid']}'),
        reason: json['reason'] as String?,
      );
    } catch (_) {
      return null;
    }
  }
}

class SiteSupervisorReply {
  const SiteSupervisorReply({
    required this.ok,
    this.state,
    this.id,
    this.pid,
    this.already = false,
    this.known = true,
    this.listening = false,
    this.processAlive = false,
    this.error,
    this.sites = const {},
  });

  final bool ok;
  final String? state;
  final String? id;
  final int? pid;
  final bool already;
  final bool known;
  final bool listening;
  final bool processAlive;
  final String? error;
  final Map<String, SiteSupervisorReply> sites;

  factory SiteSupervisorReply.fromJson(Map<String, dynamic> json) {
    final rawSites = json['sites'];
    final sites = <String, SiteSupervisorReply>{};
    if (rawSites is Map) {
      rawSites.forEach((key, value) {
        if (value is Map) {
          sites['$key'] = SiteSupervisorReply.fromJson(
            Map<String, dynamic>.from(value),
          );
        }
      });
    }
    return SiteSupervisorReply(
      ok: json['ok'] != false,
      state: json['state'] as String?,
      id: json['id'] as String?,
      pid: json['pid'] is int ? json['pid'] as int : int.tryParse('${json['pid']}'),
      already: json['already'] == true,
      known: json['known'] != false,
      listening: json['listening'] == true || json['state'] == 'listening',
      processAlive: json['processAlive'] == true,
      error: json['error'] as String?,
      sites: sites,
    );
  }

  static SiteSupervisorReply parseStdout(String stdout) {
    final line = stdout
        .trim()
        .split('\n')
        .reversed
        .map((l) => l.trim())
        .firstWhere((l) => l.startsWith('{'), orElse: () => '');
    if (line.isEmpty) {
      return const SiteSupervisorReply(ok: false, error: '看守没有返回 JSON');
    }
    try {
      final decoded = jsonDecode(line);
      if (decoded is! Map) {
        return const SiteSupervisorReply(ok: false, error: '看守返回格式错误');
      }
      return SiteSupervisorReply.fromJson(Map<String, dynamic>.from(decoded));
    } catch (e) {
      return SiteSupervisorReply(ok: false, error: '看守返回无法解析：$e');
    }
  }
}

abstract class SiteSupervisorClient {
  Stream<SiteSupervisorEvent> get events;

  Future<void> ensureReady();

  Future<SiteSupervisorReply> startSite({
    required String id,
    required String cwd,
    required String cmd,
    int? port,
    String? log,
    String? pidFile,
    Duration? timeout,
  });

  Future<SiteSupervisorReply> stopSite(String id);

  Future<SiteSupervisorReply> status(String id);

  Future<Map<String, SiteSupervisorReply>> snapshot();

  Future<void> shutdown();

  Future<void> dispose();
}

/// In-memory supervisor for tests (same protocol, no guest).
class MemorySiteSupervisorClient implements SiteSupervisorClient {
  MemorySiteSupervisorClient();

  final Map<String, SiteSupervisorReply> _sites = {};
  final StreamController<SiteSupervisorEvent> _events =
      StreamController<SiteSupervisorEvent>.broadcast();

  bool occupied = false;
  bool startFails = false;
  String startFailError = '启动超时：端口尚未监听';
  Object? throwOnStart;

  @override
  Stream<SiteSupervisorEvent> get events => _events.stream;

  void emit(SiteSupervisorEvent event) {
    if (event.down) {
      _sites[event.id] = SiteSupervisorReply(
        ok: true,
        id: event.id,
        state: event.state,
        listening: false,
        processAlive: false,
        known: true,
      );
    } else if (event.listening) {
      _sites[event.id] = SiteSupervisorReply(
        ok: true,
        id: event.id,
        state: 'listening',
        listening: true,
        processAlive: true,
        known: true,
      );
    }
    _events.add(event);
  }

  @override
  Future<void> ensureReady() async {}

  @override
  Future<SiteSupervisorReply> startSite({
    required String id,
    required String cwd,
    required String cmd,
    int? port,
    String? log,
    String? pidFile,
    Duration? timeout,
  }) async {
    final thrown = throwOnStart;
    if (thrown != null) throw thrown;
    final existing = _sites[id];
    if (existing?.listening == true) {
      return SiteSupervisorReply(
        ok: true,
        id: id,
        state: 'listening',
        already: true,
        listening: true,
        processAlive: true,
      );
    }
    if (occupied) {
      return const SiteSupervisorReply(
        ok: false,
        state: 'occupied',
        error: '端口已被其他进程占用，无法启动',
      );
    }
    if (startFails) {
      return SiteSupervisorReply(
        ok: false,
        state: 'start_failed',
        id: id,
        error: startFailError,
      );
    }
    final reply = SiteSupervisorReply(
      ok: true,
      id: id,
      state: 'listening',
      already: false,
      listening: true,
      processAlive: true,
    );
    _sites[id] = reply;
    emit(SiteSupervisorEvent(id: id, state: 'listening'));
    return reply;
  }

  @override
  Future<SiteSupervisorReply> stopSite(String id) async {
    _sites[id] = SiteSupervisorReply(
      ok: true,
      id: id,
      state: 'exited',
      listening: false,
      processAlive: false,
    );
    emit(SiteSupervisorEvent(id: id, state: 'exited'));
    return SiteSupervisorReply(ok: true, id: id, state: 'exited');
  }

  @override
  Future<SiteSupervisorReply> status(String id) async {
    return _sites[id] ??
        SiteSupervisorReply(
          ok: true,
          id: id,
          known: false,
          state: 'unknown',
          listening: false,
          processAlive: false,
        );
  }

  @override
  Future<Map<String, SiteSupervisorReply>> snapshot() async =>
      Map<String, SiteSupervisorReply>.from(_sites);

  @override
  Future<void> shutdown() async {
    _sites.clear();
  }

  @override
  Future<void> dispose() async {
    if (!_events.isClosed) {
      await _events.close();
    }
  }
}

/// Guest daemon + short RPC + dedicated `tail -F` follow.
class GuestSiteSupervisorClient implements SiteSupervisorClient {
  GuestSiteSupervisorClient(
    this.workspace, {
    this.rpcTimeout = const Duration(seconds: 25),
    Future<GuestStreamSession?> Function(String cmd)? spawnDetached,
  }) : _spawnDetached = spawnDetached ?? _spawnOnWorkspace(workspace);

  final SandboxWorkspace workspace;
  final Duration rpcTimeout;
  final Future<GuestStreamSession?> Function(String cmd) _spawnDetached;

  final StreamController<SiteSupervisorEvent> _events =
      StreamController<SiteSupervisorEvent>.broadcast();
  GuestStreamSession? _follow;
  StreamSubscription<String>? _followSub;
  bool _installed = false;

  @override
  Stream<SiteSupervisorEvent> get events => _events.stream;

  static Future<GuestStreamSession?> Function(String cmd) _spawnOnWorkspace(
    SandboxWorkspace workspace,
  ) {
    return (cmd) async {
      if (workspace is WslWorkspace) {
        return workspace.spawnDetached(cmd);
      }
      if (workspace is ProotWorkspace) {
        return workspace.spawnDetached(cmd);
      }
      return null;
    };
  }

  @override
  Future<void> ensureReady() async {
    await _install();
    final result = await workspace.run(
      siteSupervisorEnsureCommand(),
      timeout: const Duration(seconds: 12),
    );
    if (result.exitCode != 0 || !result.stdout.contains('ready')) {
      throw StateError(
        '无法启动站点看守：${result.stderr}\n${result.stdout}',
      );
    }
    await _ensureFollow();
  }

  Future<void> _install() async {
    if (_installed) return;
    await workspace.writeGuestFile(
      kSiteSupervisorPyGuestPath,
      utf8.encode(kSiteSupervisorPySource),
    );
    _installed = true;
  }

  Future<void> _ensureFollow() async {
    if (_follow != null) return;
    final session = await _spawnDetached(siteSupervisorFollowCommand());
    if (session == null) return;
    _follow = session;
    _followSub = session.lines.listen((line) {
      final event = SiteSupervisorEvent.tryParse(line);
      if (event != null && !_events.isClosed) {
        _events.add(event);
      }
    }, onDone: () {
      _follow = null;
      _followSub = null;
    });
  }

  Future<SiteSupervisorReply> _rpc(Map<String, Object?> cmd) async {
    await ensureReady();
    final result = await workspace.run(
      siteSupervisorRpcCommand(cmd),
      timeout: rpcTimeout,
    );
    if (result.exitCode != 0 && result.stdout.trim().isEmpty) {
      return SiteSupervisorReply(
        ok: false,
        error: '看守 RPC 失败：${result.stderr}\n${result.stdout}',
      );
    }
    return SiteSupervisorReply.parseStdout(result.stdout);
  }

  @override
  Future<SiteSupervisorReply> startSite({
    required String id,
    required String cwd,
    required String cmd,
    int? port,
    String? log,
    String? pidFile,
    Duration? timeout,
  }) {
    return _rpc({
      'op': 'start',
      'id': id,
      'cwd': cwd,
      'cmd': cmd,
      if (port != null) 'port': port,
      if (log != null) 'log': log,
      if (pidFile != null) 'pid_file': pidFile,
      if (timeout != null) 'timeout': timeout.inMilliseconds / 1000.0,
    });
  }

  @override
  Future<SiteSupervisorReply> stopSite(String id) =>
      _rpc({'op': 'stop', 'id': id});

  @override
  Future<SiteSupervisorReply> status(String id) =>
      _rpc({'op': 'status', 'id': id});

  @override
  Future<Map<String, SiteSupervisorReply>> snapshot() async {
    try {
      final bytes = await workspace.readGuestFile(kSiteSupervisorStateGuestPath);
      if (bytes != null && bytes.isNotEmpty) {
        final decoded = jsonDecode(utf8.decode(bytes, allowMalformed: true));
        if (decoded is Map && decoded['sites'] is Map) {
          final sites = <String, SiteSupervisorReply>{};
          (decoded['sites'] as Map).forEach((key, value) {
            if (value is Map) {
              sites['$key'] = SiteSupervisorReply.fromJson(
                Map<String, dynamic>.from(value),
              );
            }
          });
          return sites;
        }
      }
    } catch (_) {}
    final reply = await _rpc({'op': 'status'});
    return reply.sites;
  }

  @override
  Future<void> shutdown() async {
    try {
      await _rpc({'op': 'shutdown'});
    } catch (_) {}
  }

  @override
  Future<void> dispose() async {
    await _followSub?.cancel();
    _followSub = null;
    try {
      await _follow?.kill();
    } catch (_) {}
    _follow = null;
    if (!_events.isClosed) {
      await _events.close();
    }
  }
}

String siteSupervisorEnsureCommand() {
  return '''
mkdir -p ${shellSingleQuote(kGuestVaultDir)}
touch ${shellSingleQuote(kSiteSupervisorEventsGuestPath)}
pidfile=${shellSingleQuote(kSiteSupervisorPidGuestPath)}
sock=${shellSingleQuote(kSiteSupervisorSockGuestPath)}
if [ -f "\$pidfile" ]; then
  pid=\$(cat "\$pidfile" 2>/dev/null || true)
  if [ -n "\$pid" ] && [ -d "/proc/\$pid" ] && [ -S "\$sock" ]; then
    echo ready
    exit 0
  fi
fi
rm -f "\$sock"
nohup python3 ${shellSingleQuote(kSiteSupervisorPyGuestPath)} serve \\
  >>${shellSingleQuote(kSiteSupervisorLogGuestPath)} 2>&1 &
echo \$! > "\$pidfile"
i=0
while [ "\$i" -lt 50 ]; do
  if [ -S "\$sock" ]; then
    echo ready
    exit 0
  fi
  sleep 0.1
  i=\$((i+1))
done
echo not_ready
exit 1
''';
}

String siteSupervisorRpcCommand(Map<String, Object?> cmd) {
  return 'python3 ${shellSingleQuote(kSiteSupervisorPyGuestPath)} rpc ${shellSingleQuote(jsonEncode(cmd))}';
}

String siteSupervisorFollowCommand() {
  return 'tail -n0 -F ${shellSingleQuote(kSiteSupervisorEventsGuestPath)}';
}

/// Join site names for tray / notification copy (`A` or `A、B`).
String formatRunningSiteNames(Iterable<String> names) {
  return names
      .map((n) => n.trim())
      .where((n) => n.isNotEmpty)
      .join('、');
}

File: lib/agent/site_supervisor.py
#!/usr/bin/env python3
import json
import os
import signal
import socket
import subprocess
import sys
import time

VAULT = "/root/.vault"
SOCK = VAULT + "/site-supervisor.sock"
PIDFILE = VAULT + "/site-supervisor.pid"
STATE = VAULT + "/site-supervisor.state.json"
EVENTS = VAULT + "/site-supervisor.events.jsonl"
READY_DEFAULT = 15.0
POLL = 0.4

sites = {}
stop_flag = False


def ensure_dir():
    os.makedirs(VAULT, exist_ok=True)


def now():
    return int(time.time())


def emit(site_id, state, **extra):
    rec = {"v": 1, "id": site_id, "state": state, "ts": now()}
    rec.update(extra)
    line = json.dumps(rec, ensure_ascii=False) + "\n"
    with open(EVENTS, "a") as f:
        f.write(line)
        f.flush()


def write_state():
    payload = {
        "sites": {
            sid: {
                "state": s["state"],
                "pid": s.get("pid"),
                "port": s.get("port"),
                "processAlive": alive(s.get("pid")),
                "listening": s["state"] == "listening",
            }
            for sid, s in sites.items()
        }
    }
    tmp = STATE + ".tmp"
    with open(tmp, "w") as f:
        json.dump(payload, f, ensure_ascii=False)
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp, STATE)


def alive(pid):
    if not pid:
        return False
    try:
        os.kill(int(pid), 0)
        return os.path.isdir("/proc/%d" % int(pid))
    except (OSError, ValueError):
        return False


def children_of(pid):
    kids = []
    try:
        names = os.listdir("/proc")
    except OSError:
        return kids
    want = str(int(pid))
    for name in names:
        if not name.isdigit():
            continue
        try:
            with open("/proc/%s/status" % name) as f:
                for line in f:
                    if line.startswith("PPid:"):
                        if line.split()[1] == want:
                            kids.append(int(name))
                        break
        except (OSError, IndexError, ValueError):
            continue
    return kids


def listen_inodes(port):
    if not port:
        return []
    hexport = "%04X" % int(port)
    found = []
    for path in ("/proc/net/tcp", "/proc/net/tcp6"):
        try:
            with open(path) as f:
                next(f)
                for line in f:
                    parts = line.split()
                    if len(parts) < 10:
                        continue
                    local = parts[1]
                    st = parts[3].upper()
                    inode = parts[9]
                    _addr, _, p = local.rpartition(":")
                    if p.upper() == hexport and st == "0A" and inode != "0":
                        found.append(inode)
        except (OSError, StopIteration):
            continue
    return found


def pid_owns_inodes(pid, inodes):
    if not pid or not inodes:
        return False
    wanted = set(inodes)
    pids = [int(pid)] + children_of(pid)
    for p in pids:
        fd_dir = "/proc/%d/fd" % p
        try:
            names = os.listdir(fd_dir)
        except OSError:
            continue
        for name in names:
            try:
                target = os.readlink(os.path.join(fd_dir, name))
            except OSError:
                continue
            if target.startswith("socket:[") and target[8:-1] in wanted:
                return True
    return False


def port_owned_by(pid, port):
    return pid_owns_inodes(pid, listen_inodes(port))


def anyone_listening(port):
    return bool(listen_inodes(port))


def kill_tree(pid):
    if not pid or not alive(pid):
        return
    pids = children_of(pid) + [int(pid)]
    for p in pids:
        if p <= 1:
            continue
        try:
            os.kill(p, signal.SIGTERM)
        except OSError:
            pass
    time.sleep(0.35)
    for p in pids:
        if p <= 1:
            continue
        try:
            os.kill(p, signal.SIGKILL)
        except OSError:
            pass


def start_site(req):
    sid = req.get("id") or ""
    cwd = req.get("cwd") or ""
    cmd = req.get("cmd") or ""
    port = req.get("port")
    log_path = req.get("log") or ""
    pid_file = req.get("pid_file") or ""
    timeout = float(req.get("timeout") or READY_DEFAULT)
    if not sid or not cwd or not cmd:
        return {"ok": False, "error": "id/cwd/cmd 不能为空"}
    existing = sites.get(sid)
    if existing and existing.get("state") == "listening" and alive(existing.get("pid")):
        if port_owned_by(existing.get("pid"), existing.get("port") or port):
            return {
                "ok": True,
                "state": "listening",
                "id": sid,
                "pid": existing.get("pid"),
                "already": True,
                "listening": True,
                "processAlive": True,
            }
    if port and anyone_listening(port):
        owner = existing.get("pid") if existing else None
        if not owner or not port_owned_by(owner, port):
            return {
                "ok": False,
                "state": "occupied",
                "id": sid,
                "error": "端口已被其他进程占用，无法启动",
            }
    if existing and alive(existing.get("pid")):
        deadline = time.time() + timeout
        while time.time() < deadline:
            if port_owned_by(existing.get("pid"), port):
                existing["state"] = "listening"
                write_state()
                emit(sid, "listening", pid=existing.get("pid"))
                return {
                    "ok": True,
                    "state": "listening",
                    "id": sid,
                    "pid": existing.get("pid"),
                    "already": True,
                    "listening": True,
                    "processAlive": True,
                }
            if not alive(existing.get("pid")):
                break
            time.sleep(POLL)
        return {
            "ok": False,
            "state": "start_failed",
            "id": sid,
            "error": "启动超时：端口尚未监听",
        }
    os.makedirs(cwd, exist_ok=True)
    logf = None
    if log_path:
        log_dir = os.path.dirname(log_path)
        if log_dir:
            os.makedirs(log_dir, exist_ok=True)
        logf = open(log_path, "ab", buffering=0)
    try:
        proc = subprocess.Popen(
            cmd,
            cwd=cwd,
            shell=True,
            executable="/bin/sh",
            stdout=logf,
            stderr=subprocess.STDOUT,
            start_new_session=True,
        )
    except Exception as e:
        if logf:
            logf.close()
        return {"ok": False, "state": "start_failed", "id": sid, "error": str(e)}
    pid = proc.pid
    if pid_file:
        try:
            with open(pid_file, "w") as f:
                f.write("%d\n" % pid)
        except OSError:
            pass
    sites[sid] = {
        "state": "starting",
        "pid": pid,
        "port": port,
        "proc": proc,
        "log": log_path,
        "pid_file": pid_file,
    }
    write_state()
    deadline = time.time() + timeout
    while time.time() < deadline:
        if proc.poll() is not None:
            sites[sid]["state"] = "exited"
            write_state()
            emit(sid, "exited", pid=pid, code=proc.returncode)
            if logf:
                logf.close()
            return {
                "ok": False,
                "state": "start_failed",
                "id": sid,
                "error": "进程已退出",
                "pid": pid,
            }
        if port_owned_by(pid, port):
            sites[sid]["state"] = "listening"
            write_state()
            emit(sid, "listening", pid=pid)
            return {
                "ok": True,
                "state": "listening",
                "id": sid,
                "pid": pid,
                "already": False,
                "listening": True,
                "processAlive": True,
            }
        time.sleep(POLL)
    kill_tree(pid)
    sites[sid]["state"] = "exited"
    write_state()
    emit(sid, "start_failed", pid=pid)
    if logf:
        logf.close()
    return {
        "ok": False,
        "state": "start_failed",
        "id": sid,
        "error": "启动超时：端口尚未监听",
        "pid": pid,
    }


def stop_site(req):
    sid = req.get("id") or ""
    rec = sites.get(sid)
    if rec is None:
        return {"ok": True, "state": "exited", "id": sid, "already": True}
    pid = rec.get("pid")
    proc = rec.get("proc")
    kill_tree(pid)
    if proc is not None:
        try:
            proc.wait(timeout=2)
        except Exception:
            pass
    pid_file = rec.get("pid_file")
    if pid_file:
        try:
            os.remove(pid_file)
        except OSError:
            pass
    rec["state"] = "exited"
    write_state()
    emit(sid, "exited", pid=pid)
    return {"ok": True, "state": "exited", "id": sid, "already": False}


def status_site(req):
    sid = req.get("id")
    if sid:
        rec = sites.get(sid)
        if rec is None:
            return {
                "ok": True,
                "known": False,
                "id": sid,
                "listening": False,
                "processAlive": False,
                "state": "unknown",
            }
        listening = rec.get("state") == "listening" and port_owned_by(
            rec.get("pid"), rec.get("port")
        )
        return {
            "ok": True,
            "known": True,
            "id": sid,
            "state": rec.get("state"),
            "pid": rec.get("pid"),
            "listening": listening,
            "processAlive": alive(rec.get("pid")),
        }
    return {
        "ok": True,
        "sites": {
            k: {
                "state": v.get("state"),
                "pid": v.get("pid"),
                "listening": v.get("state") == "listening",
                "processAlive": alive(v.get("pid")),
            }
            for k, v in sites.items()
        },
    }


def shutdown_all():
    for sid in list(sites.keys()):
        stop_site({"id": sid})


def handle(req):
    op = (req.get("op") or "").strip()
    if op == "ping":
        return {"ok": True, "state": "pong"}
    if op == "start":
        return start_site(req)
    if op == "stop":
        return stop_site(req)
    if op == "status":
        return status_site(req)
    if op == "shutdown":
        shutdown_all()
        global stop_flag
        stop_flag = True
        return {"ok": True, "state": "shutdown"}
    return {"ok": False, "error": "未知 op"}


def monitor_once():
    for sid, rec in list(sites.items()):
        if rec.get("state") != "listening":
            continue
        pid = rec.get("pid")
        proc = rec.get("proc")
        dead = (proc is not None and proc.poll() is not None) or not alive(pid)
        if dead:
            rec["state"] = "exited"
            write_state()
            emit(sid, "exited", pid=pid)
            continue
        if rec.get("port") and not port_owned_by(pid, rec.get("port")):
            emit(sid, "port_lost", pid=pid)
            kill_tree(pid)
            rec["state"] = "exited"
            write_state()
            emit(sid, "exited", pid=pid, reason="port_lost")


def serve():
    ensure_dir()
    open(EVENTS, "a").close()
    with open(PIDFILE, "w") as f:
        f.write("%d\n" % os.getpid())
    write_state()
    try:
        os.unlink(SOCK)
    except OSError:
        pass
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.bind(SOCK)
    sock.listen(8)
    sock.settimeout(0.4)
    signal.signal(signal.SIGTERM, lambda *_: shutdown_and_exit(sock))
    signal.signal(signal.SIGINT, lambda *_: shutdown_and_exit(sock))
    while not stop_flag:
        try:
            conn, _ = sock.accept()
        except socket.timeout:
            monitor_once()
            continue
        except OSError:
            break
        try:
            data = b""
            while not data.endswith(b"\n"):
                chunk = conn.recv(4096)
                if not chunk:
                    break
                data += chunk
            req = json.loads(data.decode("utf-8") or "{}")
            reply = handle(req)
            conn.sendall((json.dumps(reply, ensure_ascii=False) + "\n").encode("utf-8"))
        except Exception as e:
            try:
                conn.sendall(
                    (json.dumps({"ok": False, "error": str(e)}) + "\n").encode("utf-8")
                )
            except OSError:
                pass
        finally:
            try:
                conn.close()
            except OSError:
                pass
        monitor_once()
    try:
        sock.close()
        os.unlink(SOCK)
    except OSError:
        pass


def shutdown_and_exit(sock):
    global stop_flag
    stop_flag = True
    shutdown_all()
    try:
        sock.close()
    except OSError:
        pass
    sys.exit(0)


def rpc(raw):
    req = json.loads(raw)
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    last = None
    for _ in range(50):
        try:
            sock.connect(SOCK)
            last = None
            break
        except OSError as e:
            last = e
            time.sleep(0.1)
    if last is not None:
        sys.stdout.write(json.dumps({"ok": False, "error": "看守未就绪：%s" % last}) + "\n")
        return 1
    sock.sendall((json.dumps(req, ensure_ascii=False) + "\n").encode("utf-8"))
    data = b""
    while not data.endswith(b"\n"):
        chunk = sock.recv(4096)
        if not chunk:
            break
        data += chunk
    sock.close()
    sys.stdout.write(data.decode("utf-8") if data else '{"ok":false,"error":"空响应"}\n')
    return 0


def main(argv):
    if len(argv) >= 2 and argv[1] == "serve":
        serve()
        return 0
    if len(argv) >= 3 and argv[1] == "rpc":
        return rpc(argv[2])
    sys.stderr.write("usage: site_supervisor.py serve | rpc '<json>'\n")
    return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv))
